lib/spack/spack/audit.py

```diff
@@ -503,6 +503,33 @@ def invalid_sha256_digest(fetcher):
     return errors
 
 
+@package_properties
+def _ensure_env_methods_are_ported_to_builders(pkgs, error_cls):
+    """Ensure that methods modifying the build environment are ported to builder classes."""
+    errors = []
+    for pkg_name in pkgs:
+        pkg_cls = spack.repo.path.get_pkg_class(pkg_name)
+        buildsystem_variant, _ = pkg_cls.variants["buildsystem"]
+        buildsystem_names = [getattr(x, "value", x) for x in buildsystem_variant.values]
+        builder_cls_names = [spack.builder.BUILDER_CLS[x].__name__ for x in buildsystem_names]
+        module = pkg_cls.module
+        has_builders_in_package_py = any(
+            getattr(module, name, False) for name in builder_cls_names
+        )
+        if not has_builders_in_package_py:
+            continue
+
+        for method_name in ("setup_build_environment", "setup_dependent_build_environment"):
```

Comment:
It occurred to me that the `cmake` package.py has a `setup_dependent_package` function with:

```
module.cmake = Executable(self.spec.prefix.bin.cmake)
```

would that be considered an issue?

* I don't think it would create conflicts: no other package should set a `.cmake` module attribute (and in general, we assume that packages will only set properties with names that would be unique to the dependency)
* We do have packages (e.g. `dray` and `ascent`) which inherit `Package` and define an `install()` method, but build with CMake (they depend on this implementation of `setup_dependent_package`)

Reply:
I don't think it creates an issue. The executable is added at module scope, so it is available both for the package and for the builder.

Reply:
That's fine: part of the goal of this PR is to resolve symbol clashes between build systems. If `setup_dependent_package` always defines non-clashing symbols then this won't be a problem. I think that's reasonable to assume and just wanted to confirm that explicitly here.